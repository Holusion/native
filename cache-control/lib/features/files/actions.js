import { INITIAL_LOAD } from "../status";

export const SET_FILES = "SET_FILES";
export const SET_DEPENDENCIES = "SET_DEPENDENCIES";
export const SET_CACHED_FILE = "SET_CACHED_FILE";
export const UNSET_CACHED_FILE = "UNSET_CACHED_FILE";


export const cacheFile = "cache_v1";

export const getFiles = (state)=>state.files.list;
export const setFiles = ({error, ...files}) =>{
  if(error){
    return {type: INITIAL_LOAD, error}
  }
  return {type: INITIAL_LOAD, files}
};


export const getDependencies = (state)=> Object.keys(state.files.list);
/**
 * 
 * @param {string="items","config"} name 
 * @param {object} list - an dict of dest->FileRef 
 */
export const setDependencies = (name, list)=>({type: SET_DEPENDENCIES, name, list});

/**
 * 
 * @param {*} state 
 * @param {string} file 
 */
export const getHash = (state, file) => getCache(state)[
  ((file.indexOf("file://") === 0) ? file.slice("file://".length) : file)
];

/**
 * 
 * @param {string} filename path to the file (absolute path is generally used)
 * @param {string} hash the file's hash. base64-encoded md5 is generally assumed
 * @returns 
 */
export const setHash = (filename, hash) => ({type:SET_CACHED_FILE, file: filename, hash});
/**
 * Remove a file from the index
 * @param {string} filename path to the file. One of the keys in `state.files.cache`
 * @returns 
 */
export const unsetHash = (filename)=>({type:UNSET_CACHED_FILE, file: filename});

/**
 * Get the whole cache state (internal use)
 * @param {*} state 
 */
export const getCache = (state)=> state.files.cache;

export const isCached = (state, name, hash)=> getCache(state)[name] === (hash || true);

/**
 * All file dependencies in cache and fresh
 * @param {import("redux").Store} state 
 * @returns {string[]} a list of file paths
 */
export const getCachedFiles = (state) => Object.keys(state.files.list).filter(f=> {
  return state.files.cache[f] && state.files.cache[f] === state.files.list[f].hash
});

/**
 * all file dependencies not in cache or stale
 * @param {import("redux").Store} state 
 * @returns {string[]}
 */
export const getUncachedFiles = (state) => Object.keys(state.files.list).filter(f=> {
  return !state.files.cache[f] || state.files.cache[f] !== state.files.list[f].hash
});


/**
 * filter out files that are not required for initial render
 * ie. videos
 * @returns {boolean}
 */
export const isRequired = (f)=> !/^video\//.test(f.contentType)

export const getRequiredFiles = (state) => getUncachedFiles(state).filter(f => isRequired(state.files.list[f]));
export const getOtherFiles = (state) => getUncachedFiles(state).filter(f=> !isRequired(state.files.list[f]));

export const getTotalSize = (state) => Object.keys(state.files.list).reduce((size, file) => (size + (state.files.list[file].size)), 0);
export const getUncachedSize = (state) => getUncachedFiles(state).reduce((size, file) => (size + (state.files.list[file].size)), 0);
export const getRequiredSize = (state)=> getRequiredFiles(state).reduce((size, file) => (size + (state.files.list[file].size)), 0);
export const getOtherSize = (state)=> getOtherFiles(state).reduce((size, file)       => (size + (state.files.list[file].size)), 0);

export const isBlocked = (state) => getRequiredFiles(state).length !== 0;