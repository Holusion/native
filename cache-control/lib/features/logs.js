'use strict';
import {SET_DATA} from "./data";
import { SET_ACTIVE_PRODUCT } from "./products";

export const ADD_LOG = "ADD_LOG";
export const LOG_INFO = "LOG_INFO";
export const LOG_WARN = "LOG_WARN";
export const LOG_ERROR = "LOG_ERROR";

export const addLine = (state, severity, {message, name, context})=> {
  let currentId = state.idCount +1;
  let errors = severity !== "error" ? state.errors: {...state.errors, [name]: currentId};
  let lines = [
    ...state.lines, 
    {id: currentId, severity, name: name, message: `${message}`, context, timestamp: new Date()}
  ];
  let slicedLines = [
    ...lines.slice(0, -100).filter(l => Object.keys(errors).findIndex((k)=>errors[k] === l.id) !== -1 ),
    ...lines.slice(-100)
  ];
  return {
    idCount: currentId,
    errors,
    lines: slicedLines,
  };
}

export default function logs(state = {
  idCount: 1,
  lines:[],  
  errors:{},
}, action) {
  if(action.error){
    return addLine(state, "error", {
      message: action.error.message,
      name: action.type,
      context: action.error.context || action.error.toString(),
    });
  }
  let clean_state = state;
  if(state.errors[action.type]){
    let {[action.type]:_, ...errors} = state.errors;
    clean_state = {...state, errors};
  }
  switch(action.type){
    case LOG_INFO:
      return addLine(clean_state , "info", action);
    case LOG_WARN:
      return addLine(clean_state , "warn", action);
    case LOG_ERROR:
      return addLine(clean_state , "error", action);
    case SET_DATA:
      if(action.data["items"]){
        return addLine(clean_state, "info", {message: "Mise à jour des pages", name: SET_DATA, context: `Réception de ${Object.keys(action.data["items"]).length} pages` })
      }else{
        return addLine(clean_state, "info", {message: "Mise à jour du projet", name: SET_DATA, context: `Configuration : ${JSON.stringify(action.data, null, 2)}` })
      }
    case SET_ACTIVE_PRODUCT:
      return addLine(clean_state, "info", {message: `Connexion à ${action.name}`, name:SET_ACTIVE_PRODUCT});
    default:
      return clean_state ;
  }
}


export const getLogs = (state)=> state.logs.lines;

export const getErrors = (state)=> {
  let lines = getLogs(state);
  return Object.keys(state.logs.errors).map(name => lines.filter(l=>l.id === state.logs.errors[name])).flat();
}

export const getError = (state, name)=>{
  return getErrors(state).find(e=> e.name === name);
}

/**
 * @typedef {object} LogAction
 * @property {typeof LOG_INFO|typeof LOG_WARN|typeof LOG_ERROR} type
 * @property {string} name generally the type of the action that called the event is used
 * @property {string} message
 * @property {string} [context]
 */

/**
 * 
 * @returns {LogAction}
 */
export const log = (severity, nameOrMessage, message, context)=>{
  return {type:severity, name: message?nameOrMessage: severity.split("_").slice(-1)[0].toUpperCase(), message:message? message:nameOrMessage, context};
}
/**
 * 
 * @param  {string} [name] A custom log name
 * @param {string} message - the log's main message
 * @param {string} [context] - optionnal extended context
 */
export const info = (...args)=> log(LOG_INFO, ...args);
export const warn = (...args)=> log(LOG_WARN, ...args);
export const error = (...args)=> log(LOG_ERROR, ...args);
