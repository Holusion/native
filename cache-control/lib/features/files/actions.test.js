'use strict';

import {reducers} from "..";
import {
  setDependencies, getDependencies, 
  setFiles, getUncachedFiles, getRequiredFiles, getOtherFiles, SET_DEPENDENCIES
} from ".";

import {makeFileRef} from "./_mock_fileRef";
import { getRequiredSize, getOtherSize, SET_CACHED_FILE, isCached, getCachedFiles, getHash, setHash, unsetHash } from './actions';
import { getError } from '../logs';


describe("files reducer", function () {
  let initialState;
  beforeEach(()=>{
    initialState = reducers(reducers(undefined, {}), setFiles({
      list: {
        "/tmp/foo.png": makeFileRef("foo.png", {size: 32}),
        "/tmp/foofoo.png": makeFileRef("foofoo.png", {size: 64}),
        "/tmp/bar.mp4": makeFileRef("bar.mp4", {size: 128}),
        "/tmp/barbar.mp4": makeFileRef("barbar.mp4", {size: 256}),
      },
      itemFiles:["/tmp/foo.png", "/tmp/foofoo.png"],
      configFiles: ["/tmp/bar.mp4", "/tmp/barbar.mp4"],
      cache: {"/tmp/foofoo.png": "xxxxxx", "/tmp/barbar.mp4":"xxxxxx"},
    }));
  })
  describe("setFiles()", ()=>{
    test("initial state shape", ()=>{
      expect(getDependencies(initialState)).toHaveLength(4);
      expect(initialState).toMatchSnapshot();
    })
    test("ignore errors", ()=>{
      expect(reducers(initialState, setFiles({error: new Error("Booh")}))).toHaveProperty("files", initialState.files);
    })
  });

  describe("setDependencies()", ()=>{
    test("change files dependencies for items", ()=>{
      let s = reducers(initialState, setDependencies("items", {
        "/tmp/baz.png": makeFileRef("baz.png"),
      }));
      expect(getDependencies(s)).toHaveLength(3);
      expect(s.files).toMatchSnapshot();
    });
    test("change files dependencies for config", ()=>{
      let s = reducers(initialState, setDependencies("config", {
        "/tmp/baz.png": makeFileRef("baz.png"),
      }));
      expect(getDependencies(s)).toHaveLength(3);
      expect(s.files).toMatchSnapshot();
    });

    test("merge dependencies for items and config", ()=>{
      let s = reducers(undefined, setDependencies("items", {
        "/tmp/baz.png": makeFileRef("baz.png"),
      }));
      s = reducers(s, setDependencies("config", {
        "/tmp/baz.png": makeFileRef("baz.png"),
      }));
      expect(getDependencies(s)).toHaveLength(1);
      expect(s.files).toMatchSnapshot();
    })

    test("remove unused dependencies", ()=>{
      let s = reducers(initialState, setDependencies("items", {}));
      expect(getDependencies(s)).toHaveLength(2);
      expect(s.files).toMatchSnapshot();
    })

    test("update file hash", ()=>{
      let s = reducers(initialState, setDependencies("items", {
        "/tmp/foo.png": makeFileRef("foo.png", {hash:"aaaaaa"}),
        "/tmp/foofoo.npg": makeFileRef("foofoo.png"),
      }));
      expect(getDependencies(s)).toHaveLength(4);
      expect(s.files).toMatchSnapshot();
    })

    test("handle dependencies errors", ()=>{
      let err = new Error("Booh");
      let s = reducers(initialState, {type: SET_DEPENDENCIES, error: err});
      expect(s.files).toBe(initialState.files);
      expect(getError(s, SET_DEPENDENCIES)).toEqual(expect.objectContaining({
        name: SET_DEPENDENCIES,
        message: err.message,
        severity: "error"
      }));
    })

    test("handle setHash errors", ()=>{
      let err = new Error("Booh");
      let s = reducers(initialState, {type: SET_CACHED_FILE, error: err});
      expect(s.files).toBe(initialState.files);
      expect(getError(s, SET_CACHED_FILE)).toEqual(expect.objectContaining({
        name: SET_CACHED_FILE,
        message: err.message,
        severity: "error"
      }));
    })
  })

  describe("setHash() / getHash()", ()=>{
    test("uses cached files list", ()=>{
      expect(getHash(initialState, "/tmp/foofoo.png")).toEqual("xxxxxx");
    })
    test("returns undefined for uncached files", ()=>{
      expect(getHash(initialState, "/tmp/foo.png")).toBeUndefined();
    })
    test("also match file:// prefix", ()=>{
      expect(getHash(initialState, "file:///tmp/foofoo.png")).toEqual("xxxxxx");
    })

    test("set a file's hash", ()=>{
      let s = reducers(initialState, setHash("/tmp/foo/bar.mp4", "spYioUcdUZuOcP2/N8eCkQ=="));
      expect(getHash(s, "/tmp/foo/bar.mp4")).toEqual("spYioUcdUZuOcP2/N8eCkQ==");
    });
  })

  describe("unsetHash()", ()=>{
    test("removes a file from cache", ()=>{
      expect(getHash(initialState, "/tmp/foofoo.png")).toEqual("xxxxxx");
      let s = reducers(initialState, unsetHash("/tmp/foofoo.png"));
      expect(s.files.cache).not.toEqual(initialState.files.cache);
      expect(getHash(s, "/tmp/foofoo.png")).toBeUndefined();
    });

    test("don't throw if file is absent", ()=>{
      let cache = initialState.files.cache;
      let s = reducers(initialState, unsetHash("/tmp/absent-file.png"));
      expect(s.files.cache).toEqual(cache);
    });
  })


  describe("getUncachedFiles()", ()=>{
    test("list files not in cache", ()=>{
      expect(getUncachedFiles(initialState)).toHaveLength(2);
    });
    test("match files that have incorrect cached hash", ()=>{
      let s = reducers(undefined, setFiles({
        list: {"/tmp/foo.png": makeFileRef("foo.png")},
        itemFiles: ["/tmp/foo.png"],
        cache: {"/tmp/foo.png":"aaaaaa"}
      }));
      expect(getUncachedFiles(s)).toHaveLength(1);
    })
    test("don't match cached files", ()=>{
      let s = reducers(initialState, setFiles({
        list: {"/tmp/foo.png": makeFileRef("foo.png")},
        itemFiles: ["/tmp/foo.png"],
        cache: {"/tmp/foo.png":"xxxxxx"}
      }));
      expect(getUncachedFiles(s)).toHaveLength(0);
    })
  })

  describe("getCachedFiles()", ()=>{
    test("list files in cache", ()=>{
      expect(getCachedFiles(initialState)).toMatchSnapshot();
    })
  })
  describe("getRequiredFiles()", ()=>{
    test("don't match video files or cached files", ()=>{
      let files = getRequiredFiles(initialState);
      expect(files).toEqual(["/tmp/foo.png"]);
    })
  })

  describe("getOtherFiles()", ()=>{
    test("don't match video files", ()=>{
      let files = getOtherFiles(initialState);
      expect(files).toHaveLength(1);
      expect(files).toEqual(expect.arrayContaining(["/tmp/bar.mp4"]));
    })
  })


  describe("getRequiredSize()", ()=>{
    test("don't match video files", ()=>{
      let size = getRequiredSize(initialState);
      expect(size).toEqual(32);
    });
  })

  describe("getOtherSize()", ()=>{
    test("don't match video files", ()=>{
      let size = getOtherSize(initialState);
      expect(size).toEqual(128);
    });
  })

  describe("isCached()", ()=>{
    test("always returns false if hash is not provided", ()=>{
      expect(isCached({files: {cache:{"/tmp/foo.png":"xxxxxx"}}}, "/tmp/foo.png", null)).toBe(false);
    })
    test("returns true if hash matches cache", ()=>{
      expect(isCached({files: {cache:{"/tmp/foo.png":"xxxxxx"}}}, "/tmp/foo.png", "xxxxxx")).toBe(true);
    })
  })
})
