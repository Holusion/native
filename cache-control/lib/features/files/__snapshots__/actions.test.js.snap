// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`files reducer getCachedFiles() list files in cache 1`] = `
Array [
  "/tmp/foofoo.png",
  "/tmp/barbar.mp4",
]
`;

exports[`files reducer setDependencies() change files dependencies for config 1`] = `
Object {
  "cache": Object {
    "/tmp/barbar.mp4": "xxxxxx",
    "/tmp/foofoo.png": "xxxxxx",
  },
  "configFiles": Array [
    "/tmp/baz.png",
  ],
  "itemFiles": Array [
    "/tmp/foo.png",
    "/tmp/foofoo.png",
  ],
  "list": Object {
    "/tmp/baz.png": Object {
      "contentType": "image/png",
      "hash": "xxxxxx",
      "name": "baz.png",
      "size": 16,
      "src": "gs://example.com/baz.png",
    },
    "/tmp/foo.png": Object {
      "contentType": "image/png",
      "hash": "xxxxxx",
      "name": "foo.png",
      "size": 32,
      "src": "gs://example.com/foo.png",
    },
    "/tmp/foofoo.png": Object {
      "contentType": "image/png",
      "hash": "xxxxxx",
      "name": "foofoo.png",
      "size": 64,
      "src": "gs://example.com/foofoo.png",
    },
  },
}
`;

exports[`files reducer setDependencies() change files dependencies for items 1`] = `
Object {
  "cache": Object {
    "/tmp/barbar.mp4": "xxxxxx",
    "/tmp/foofoo.png": "xxxxxx",
  },
  "configFiles": Array [
    "/tmp/bar.mp4",
    "/tmp/barbar.mp4",
  ],
  "itemFiles": Array [
    "/tmp/baz.png",
  ],
  "list": Object {
    "/tmp/bar.mp4": Object {
      "contentType": "video/mp4",
      "hash": "xxxxxx",
      "name": "bar.mp4",
      "size": 128,
      "src": "gs://example.com/bar.mp4",
    },
    "/tmp/barbar.mp4": Object {
      "contentType": "video/mp4",
      "hash": "xxxxxx",
      "name": "barbar.mp4",
      "size": 256,
      "src": "gs://example.com/barbar.mp4",
    },
    "/tmp/baz.png": Object {
      "contentType": "image/png",
      "hash": "xxxxxx",
      "name": "baz.png",
      "size": 16,
      "src": "gs://example.com/baz.png",
    },
  },
}
`;

exports[`files reducer setDependencies() merge dependencies for items and config 1`] = `
Object {
  "cache": Object {},
  "configFiles": Array [
    "/tmp/baz.png",
  ],
  "itemFiles": Array [
    "/tmp/baz.png",
  ],
  "list": Object {
    "/tmp/baz.png": Object {
      "contentType": "image/png",
      "hash": "xxxxxx",
      "name": "baz.png",
      "size": 16,
      "src": "gs://example.com/baz.png",
    },
  },
}
`;

exports[`files reducer setDependencies() remove unused dependencies 1`] = `
Object {
  "cache": Object {
    "/tmp/barbar.mp4": "xxxxxx",
    "/tmp/foofoo.png": "xxxxxx",
  },
  "configFiles": Array [
    "/tmp/bar.mp4",
    "/tmp/barbar.mp4",
  ],
  "itemFiles": Array [],
  "list": Object {
    "/tmp/bar.mp4": Object {
      "contentType": "video/mp4",
      "hash": "xxxxxx",
      "name": "bar.mp4",
      "size": 128,
      "src": "gs://example.com/bar.mp4",
    },
    "/tmp/barbar.mp4": Object {
      "contentType": "video/mp4",
      "hash": "xxxxxx",
      "name": "barbar.mp4",
      "size": 256,
      "src": "gs://example.com/barbar.mp4",
    },
  },
}
`;

exports[`files reducer setDependencies() update file hash 1`] = `
Object {
  "cache": Object {
    "/tmp/barbar.mp4": "xxxxxx",
    "/tmp/foofoo.png": "xxxxxx",
  },
  "configFiles": Array [
    "/tmp/bar.mp4",
    "/tmp/barbar.mp4",
  ],
  "itemFiles": Array [
    "/tmp/foo.png",
    "/tmp/foofoo.npg",
  ],
  "list": Object {
    "/tmp/bar.mp4": Object {
      "contentType": "video/mp4",
      "hash": "xxxxxx",
      "name": "bar.mp4",
      "size": 128,
      "src": "gs://example.com/bar.mp4",
    },
    "/tmp/barbar.mp4": Object {
      "contentType": "video/mp4",
      "hash": "xxxxxx",
      "name": "barbar.mp4",
      "size": 256,
      "src": "gs://example.com/barbar.mp4",
    },
    "/tmp/foo.png": Object {
      "contentType": "image/png",
      "hash": "aaaaaa",
      "name": "foo.png",
      "size": 16,
      "src": "gs://example.com/foo.png",
    },
    "/tmp/foofoo.npg": Object {
      "contentType": "image/png",
      "hash": "xxxxxx",
      "name": "foofoo.png",
      "size": 16,
      "src": "gs://example.com/foofoo.png",
    },
  },
}
`;

exports[`files reducer setFiles() initial state shape 1`] = `
Object {
  "conf": Object {
    "configurableProjectName": true,
    "default_target": null,
    "play_control": "none",
    "projectName": undefined,
    "purge_products": false,
    "slides_control": "default",
    "timeout": 0,
    "watch": true,
  },
  "data": Object {
    "config": Object {},
    "items": Object {},
  },
  "files": Object {
    "cache": Object {
      "/tmp/barbar.mp4": "xxxxxx",
      "/tmp/foofoo.png": "xxxxxx",
    },
    "configFiles": Array [
      "/tmp/bar.mp4",
      "/tmp/barbar.mp4",
    ],
    "itemFiles": Array [
      "/tmp/foo.png",
      "/tmp/foofoo.png",
    ],
    "list": Object {
      "/tmp/bar.mp4": Object {
        "contentType": "video/mp4",
        "hash": "xxxxxx",
        "name": "bar.mp4",
        "size": 128,
        "src": "gs://example.com/bar.mp4",
      },
      "/tmp/barbar.mp4": Object {
        "contentType": "video/mp4",
        "hash": "xxxxxx",
        "name": "barbar.mp4",
        "size": 256,
        "src": "gs://example.com/barbar.mp4",
      },
      "/tmp/foo.png": Object {
        "contentType": "image/png",
        "hash": "xxxxxx",
        "name": "foo.png",
        "size": 32,
        "src": "gs://example.com/foo.png",
      },
      "/tmp/foofoo.png": Object {
        "contentType": "image/png",
        "hash": "xxxxxx",
        "name": "foofoo.png",
        "size": 64,
        "src": "gs://example.com/foofoo.png",
      },
    },
  },
  "logs": Object {
    "errors": Object {},
    "idCount": 1,
    "lines": Array [],
  },
  "products": Array [],
  "status": Object {
    "connected": false,
    "initial_load": true,
    "signedIn": false,
    "synchronized": false,
  },
}
`;
